// @webex/internal-plugin-device/dist/index.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");

var _interopRequireWildcard = require("@babel/runtime-corejs2/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

_Object$defineProperty(exports, "default", {
  enumerable: true,
  get: function get() {
    return _device.default;
  }
});

_Object$defineProperty(exports, "FeatureCollection", {
  enumerable: true,
  get: function get() {
    return _index.FeatureCollection;
  }
});

_Object$defineProperty(exports, "FeatureModel", {
  enumerable: true,
  get: function get() {
    return _index.FeatureModel;
  }
});

_Object$defineProperty(exports, "FeaturesModel", {
  enumerable: true,
  get: function get() {
    return _index.FeaturesModel;
  }
});

_Object$defineProperty(exports, "DeviceUrlInterceptor", {
  enumerable: true,
  get: function get() {
    return _deviceUrl.default;
  }
});

_Object$defineProperty(exports, "config", {
  enumerable: true,
  get: function get() {
    return _config.default;
  }
});

exports.constants = void 0;

var _webexCore = require("@webex/webex-core");

var _device = _interopRequireDefault(require("./device"));

var _index = require("./features/index");

var _deviceUrl = _interopRequireDefault(require("./interceptors/device-url"));

var constants = _interopRequireWildcard(require("./constants"));

exports.constants = constants;

var _config = _interopRequireDefault(require("./config"));

// Internal dependencies.
// Plugin dependencies.
(0, _webexCore.registerInternalPlugin)('device', _device.default, {
  config: _config.default,
  interceptors: {
    DeviceUrlInterceptor: _deviceUrl.default.create
  },

  /**
   * Unregister the device in the case that the webex instance has logged out.
   *
   * @returns {Promise<undefined>}
   */
  onBeforeLogout: function onBeforeLogout() {
    return this.unregister();
  }
});
