// @webex/webex-core/dist/lib/storage/index.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");

var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

_Object$defineProperty(exports, "persist", {
  enumerable: true,
  get: function get() {
    return _decorators.persist;
  }
});

_Object$defineProperty(exports, "waitForValue", {
  enumerable: true,
  get: function get() {
    return _decorators.waitForValue;
  }
});

_Object$defineProperty(exports, "makeWebexStore", {
  enumerable: true,
  get: function get() {
    return _makeWebexStore.default;
  }
});

_Object$defineProperty(exports, "makeWebexPluginStore", {
  enumerable: true,
  get: function get() {
    return _makeWebexPluginStore.default;
  }
});

_Object$defineProperty(exports, "MemoryStoreAdapter", {
  enumerable: true,
  get: function get() {
    return _memoryStoreAdapter.default;
  }
});

_Object$defineProperty(exports, "StorageError", {
  enumerable: true,
  get: function get() {
    return _errors.StorageError;
  }
});

_Object$defineProperty(exports, "NotFoundError", {
  enumerable: true,
  get: function get() {
    return _errors.NotFoundError;
  }
});

var _decorators = require("./decorators");

var _makeWebexStore = _interopRequireDefault(require("./make-webex-store.js"));

var _makeWebexPluginStore = _interopRequireDefault(require("./make-webex-plugin-store.js"));

var _memoryStoreAdapter = _interopRequireDefault(require("./memory-store-adapter"));

var _errors = require("./errors");
