// @webex/helper-html/dist/html.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");

var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

_Object$defineProperty(exports, "escape", {
  enumerable: true,
  get: function get() {
    return _htmlBase.escape;
  }
});

_Object$defineProperty(exports, "escapeSync", {
  enumerable: true,
  get: function get() {
    return _htmlBase.escapeSync;
  }
});

exports.filterEscapeSync = exports.filterEscape = exports.filterSync = exports.filter = void 0;

var _promise = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/promise"));

var _curry2 = _interopRequireDefault(require("lodash/curry"));

var _htmlBase = require("./html-base");

/**
 * @param {Object} allowedTags
 * @param {Array<string>} allowedStyles
 * @param {string} html
 * @private
 * @returns {string}
 */
function noop() {
  for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
    args[_key] = arguments[_key];
  }

  return new _promise.default(function (resolve) {
    resolve(noopSync.apply(void 0, args));
  });
}
/**
 * @param {function} processCallback callback function to do additional
 * processing on node. of the form process(node)
 * @param {Object} allowedTags
 * @param {Array<string>} allowedStyles
 * @param {string} html
 * @private
 * @returns {string}
 */


function noopSync(processCallback, allowedTags, allowedStyles, html) {
  return html;
}

var filter = (0, _curry2.default)(noop, 4);
exports.filter = filter;
var filterSync = (0, _curry2.default)(noopSync, 4);
exports.filterSync = filterSync;
var filterEscape = (0, _curry2.default)(noop, 4);
exports.filterEscape = filterEscape;
var filterEscapeSync = (0, _curry2.default)(noopSync, 4);
exports.filterEscapeSync = filterEscapeSync;
